routing/index_graph_starter.cpp

```diff
@@ -1,180 +1,396 @@
+  for (auto const & kv : m_fake.m_fakeToReal)
+    mwms.insert(kv.second.GetMwmId());
 
-// static
-vector<Segment>::const_iterator IndexGraphStarter::GetNonFakeFinishIt(vector<Segment> const & segments)
-{
-  CheckValidRoute(segments);
-  // See CheckValidRoute comment.
-  return segments.end() - 2;
+  return mwms;
 }
 
 // static
 size_t IndexGraphStarter::GetRouteNumPoints(vector<Segment> const & segments)
 {
   CheckValidRoute(segments);
-  return segments.size() - 1;
+  return segments.size() + 1;
 }
 
 Junction const & IndexGraphStarter::GetRouteJunction(vector<Segment> const & segments,
-                                                     size_t pointIndex)
+                                                     size_t pointIndex) const
 {
-  if (pointIndex == 0)
-    return m_start.GetJunction();
+  CHECK(!segments.empty(), ());
+  CHECK_LESS_OR_EQUAL(
+      pointIndex, segments.size(),
+      ("Point with index", pointIndex, "does not exist in route with size", segments.size()));
+  if (pointIndex == segments.size())
+    return GetJunction(segments[pointIndex - 1], true);
+  return GetJunction(segments[pointIndex], false);
+}
+
+double IndexGraphStarter::CalcSegmentWeight(Segment const & segment) const
+{
+  if (!IsFakeSegment(segment))
+  {
+    return m_graph.GetEstimator().CalcSegmentWeight(
+        segment, m_graph.GetRoadGeometry(segment.GetMwmId(), segment.GetFeatureId()));
+  }
 
-  if (pointIndex + 1 == GetRouteNumPoints(segments))
-    return m_finish.GetJunction();
+  auto const fakeVertexIt = m_fake.m_segmentToVertex.find(segment);
 
-  CHECK_LESS(pointIndex, segments.size(), ());
-  return GetJunction(segments[pointIndex], true /* front */);
+  CHECK(fakeVertexIt != m_fake.m_segmentToVertex.end(),
+        ("Requested junction for invalid fake segment."));
+  return m_graph.GetEstimator().CalcLeapWeight(fakeVertexIt->second.GetPointFrom(),
+                                               fakeVertexIt->second.GetPointTo());
 }
 
-void IndexGraphStarter::GetEdgesList(Segment const & segment, bool isOutgoing,
-                                     vector<SegmentEdge> & edges)
+double IndexGraphStarter::CalcRouteSegmentWeight(vector<Segment> const & route,
+                                                 size_t segmentIndex) const
 {
-  edges.clear();
+  CHECK(!route.empty(), ());
+  CHECK_LESS(
+      segmentIndex, route.size(),
+      ("Segment with index", segmentIndex, "does not exist in route with size", route.size()));
+  return CalcSegmentWeight(route[segmentIndex]);
+}
 
-  if (segment == kStartFakeSegment)
+bool IndexGraphStarter::IsLeap(NumMwmId mwmId) const
+{
+  bool inProjectedMwmIds = false;
+  for (auto const & kv : m_fake.m_fakeToReal)
   {
-    GetFakeToNormalEdges(m_start, isOutgoing, edges);
-    return;
+    if (kv.second.GetMwmId() == mwmId)
+    {
+      inProjectedMwmIds = true;
+      break;
+    }
   }
 
-  if (segment == kFinishFakeSegment)
+  return mwmId != kFakeNumMwmId && !inProjectedMwmIds &&
```

Comment:
I don't know is there a performance issue, but is it possible to move out the first clause for early return? I.e.:
```c++
if (mwmId == kFakeNumMwmId)
  return false;

bool inProjectedMwmIds = false;
...
return !inProjectedMwmIds && ...
```
?